test/k8sT/istio.go

```diff
@@ -0,0 +1,276 @@
+			res := kubectl.ExecPodCmd(
+				helpers.DefaultNamespace, srcPod, fmt.Sprintf("%s %s", wgetCommand, dstURI))
+			if res.WasSuccessful() {
+				GinkgoPrint("Was able to connect from %q to %q, but expected no connection: %s", srcPod, dstURI, res.CombineOutput())
+				return false
+			}
+			return true
+		}
+
+		// formatLabelArgument formats the provided key-value pairs as labels for use in
+		// querying Kubernetes.
+		formatLabelArgument := func(firstKey, firstValue string, nextLabels ...string) string {
+			baseString := fmt.Sprintf("-l %s=%s", firstKey, firstValue)
+			if nextLabels == nil {
+				return baseString
+			} else if len(nextLabels)%2 != 0 {
+				panic("must provide even number of arguments for label key-value pairings")
+			} else {
+				for i := 0; i < len(nextLabels); i += 2 {
+					baseString = fmt.Sprintf("%s,%s=%s", baseString, nextLabels[i], nextLabels[i+1])
+				}
+			}
+			return baseString
+		}
+
+		// formatAPI is a helper function which formats a URI to access.
+		formatAPI := func(service, port, resource string) string {
+			target := fmt.Sprintf(
+				"%s.%s.svc.cluster.local:%s",
+				service, helpers.DefaultNamespace, port)
+			if resource != "" {
+				return fmt.Sprintf("%s/%s", target, resource)
+			}
+			return target
+		}
+
+		It("Tests bookinfo inter-service connectivity", func() {
+			version := "version"
+			v1 := "v1"
+
+			productPage := "productpage"
+			reviews := "reviews"
+			ratings := "ratings"
+			details := "details"
+			dnsChecks := []string{productPage, reviews, ratings, details}
+			app := "app"
+			health := "health"
+			ratingsPath := "ratings/0"
+
+			apiPort := "9080"
+
+			podNameFilter := "{.items[*].metadata.name}"
+
+			By("Waiting for pods to be ready")
+			err := kubectl.WaitforPods(helpers.DefaultNamespace, "-l zgroup=bookinfo", helpers.HelperTimeout)
+			Expect(err).Should(BeNil(), "Pods are not ready after timeout")
+
+			By("Waiting for services to be ready")
+			for _, service := range []string{details, ratings, reviews, productPage} {
+				err = kubectl.WaitForServiceEndpoints(
+					helpers.DefaultNamespace, "", service,
+					apiPort, helpers.HelperTimeout)
+				Expect(err).Should(BeNil(), "Service %q is not ready after timeout", service)
+			}
+
+			By("Validating DNS resolution")
+			for _, name := range dnsChecks {
+				err = kubectl.WaitForKubeDNSEntry(name, helpers.DefaultNamespace)
+				Expect(err).To(BeNil(), "DNS entry is not ready after timeout")
+			}
+
+			By("Validating L7 filtering")
+			reviewsPodV1, err := kubectl.GetPods(helpers.DefaultNamespace, formatLabelArgument(app, reviews, version, v1)).Filter(podNameFilter)
+			Expect(err).Should(BeNil(), "Cannot get reviewsV1 pods")
+			productpagePodV1, err := kubectl.GetPods(helpers.DefaultNamespace, formatLabelArgument(app, productPage, version, v1)).Filter(podNameFilter)
+			Expect(err).Should(BeNil(), "Cannot get productpageV1 pods")
+
+			// Connectivity checks often need to be repeated because Pilot
```

Comment:
Could we wailt until the sidecar is ready? is there any way to know that?

Reply:
It may be possible to use the new `istioctl` debug commands. But I don't want to rely on that. Testing connectivity and checking that we see no 503s is the real test.

Reply:
Hi!

The thing that I want to avoid is to wait until things are ready, and don't try
to connect when things are ok, for me, if a proxy stop to return 503, and start
to work, the next connection should be ok, so I'll keep the `withTimeout` and
after that, validate that all connection works, so we're use that connections works more than once.


Added something like this with ExpectSuccess(), after the withTimeout should be ok:
```
shouldConnect(reviewsPodV1.String(), formatAPI(ratings, apiPort, health))
shouldNotConnect(reviewsPodV1.String(), formatAPI(ratings, apiPort, ratingsPath))
shouldConnect(productpagePodV1.String(), formatAPI(details, apiPort, health))
shouldNotConnect(productpagePodV1.String(), formatAPI(ratings, apiPort, health))
shouldNotConnect(productpagePodV1.String(), formatAPI(ratings, apiPort, ratingsPath))
```

Reply:
OK, if you are happy with that, I'm happy then :-) wanted to make sure that no weird cases happens ;-)